File: avalon/platform/android-google/gamecenter/GameCenter.cpp
#include <avalon/GameCenter.h>

#include <jni.h>
#include "cocos2d.h"
#include "platform/android/jni/JniHelper.h"

namespace avalon {

/**
 * C++ -->> Java
 */

namespace helper {
namespace gamecenter {

const char* const CLASS_NAME = "com/avalon/GameCenter";

void callStaticVoidMethod(const char* name)
{
    cocos2d::JniMethodInfo t;
    if (cocos2d::JniHelper::getStaticMethodInfo(t, CLASS_NAME, name, "()V")) {
        t.env->CallStaticVoidMethod(t.classID, t.methodID);
        t.env->DeleteLocalRef(t.classID);
    }
}

bool callStaticBoolMethod(const char* name)
{
    cocos2d::JniMethodInfo t;
    if (cocos2d::JniHelper::getStaticMethodInfo(t, CLASS_NAME, name, "()Z")) {
        bool result = (t.env->CallStaticBooleanMethod(t.classID, t.methodID) == JNI_TRUE);
        t.env->DeleteLocalRef(t.classID);
        return result;
    } else {
        return false;
    }
}

void callStaticVoidMethodWithString(const char* name, const char* idName)
{
    cocos2d::JniMethodInfo t;
    if (cocos2d::JniHelper::getStaticMethodInfo(t, CLASS_NAME, name, "(Ljava/lang/String;)V")) {
        jstring jIdName = t.env->NewStringUTF(idName);
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jIdName);
        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jIdName);
    }
}

void callStaticVoidMethodWithStringAndInt(const char* name, const char* idName, const int score)
{
    cocos2d::JniMethodInfo t;
    if (cocos2d::JniHelper::getStaticMethodInfo(t, CLASS_NAME, name, "(Ljava/lang/String;I)V")) {
        jstring jIdName = t.env->NewStringUTF(idName);
        t.env->CallStaticVoidMethod(t.classID, t.methodID, jIdName, (jint)score);
        t.env->DeleteLocalRef(t.classID);
        t.env->DeleteLocalRef(jIdName);
    }
}

} // namespace gamecenter
} // namespace helper

/**
 * Public API
 */

void GameCenter::login()
{
    helper::gamecenter::callStaticVoidMethod("login");
}

bool GameCenter::showAchievements()
{
    return helper::gamecenter::callStaticBoolMethod("showAchievements");
}

void GameCenter::postAchievement(const char* idName, int percentComplete)
{
    helper::gamecenter::callStaticVoidMethodWithStringAndInt("postAchievement", idName, percentComplete);
}

void GameCenter::clearAllAchievements()
{
    helper::gamecenter::callStaticVoidMethod("clearAllAchievements");
}

bool GameCenter::showScores()
{
    return helper::gamecenter::callStaticBoolMethod("showScores");
}

void GameCenter::postScore(const char* idName, int score)
{
    helper::gamecenter::callStaticVoidMethodWithStringAndInt("postScore", idName, score);
}

void GameCenter::clearAllScores()
{
    helper::gamecenter::callStaticVoidMethod("clearAllScores");
}

void GameCenter::signIn()
{
    helper::gamecenter::callStaticVoidMethod("signIn");
}

void GameCenter::signOut()
{
    helper::gamecenter::callStaticVoidMethod("signOut");
}

bool GameCenter::isSignedIn()
{
    return helper::gamecenter::callStaticBoolMethod("isSignedIn");

}

} // namespace avalon

File: avalon/platform/android-google/gamecenter/java/src/com/avalon/GameCenter.java
package com.avalon;

import android.util.Log;
import android.content.Intent;

import org.cocos2dx.lib.Cocos2dxActivity;
import com.google.example.games.basegameutils.GameHelper;

public abstract class GameCenter
{
    private static GameHelper mHelper = null;
    private static final String TAG = "avalon.GameCenter";
    private static final Cocos2dxActivity activity = (Cocos2dxActivity) Cocos2dxActivity.getContext();
    private final static int REQUEST_ACHIEVEMENTS = 9003;
    private final static int REQUEST_LEADERBOARD = 9004;

    /**
     *
     * Interface used for C++
     *
     */
     
    public static void login()
    {
    }

    public static void signIn()
    {
        activity.runOnUiThread(new Runnable() {
            public void run() {
                getGameHelper().beginUserInitiatedSignIn();
            }
        });
    }

    public static void signOut()
    {
        if (mHelper == null) {
            return;
        }

        activity.runOnUiThread(new Runnable() {
            public void run() {
                getGameHelper().signOut();
            }
        });
    }

    public static boolean isSignedIn()
    {
        return (mHelper != null && getGameHelper().isSignedIn());
    }

    public static boolean showAchievements()
    {
        if (!isSignedIn()) {
            return false;
        }
        
        activity.runOnUiThread(new Runnable() {
            public void run() {
                activity.startActivityForResult(
                    getGameHelper().getGamesClient().getAchievementsIntent(),
                    REQUEST_ACHIEVEMENTS
                );
            }
        });
        return true;
    }

    public static void postAchievement(final String idName, int percentComplete)
    {
        if (!isSignedIn() || percentComplete < 100) {
            return;
        }

        activity.runOnUiThread(new Runnable() {
            public void run() {
                int rId = getResourceId(idName);
                if (rId > 0) {
                    String googleId = activity.getResources().getString(rId);
                    getGameHelper().getGamesClient().unlockAchievement(googleId);
                }
            }
        });
    }

    public static void clearAllAchievements()
    {
        Log.v(TAG, "clearAllAchievements is not available on this platform");
    }

    public static boolean showScores()
    {
        if (!isSignedIn()) {
            return false;
        }
        
        activity.runOnUiThread(new Runnable() {
            public void run() {
                activity.startActivityForResult(
                    getGameHelper().getGamesClient().getAllLeaderboardsIntent(),
                    REQUEST_LEADERBOARD
                );
            }
        });
        return true;
    }

    public static void postScore(final String idName, final int score)
    {
        if (!isSignedIn()) {
            return;
        }

        activity.runOnUiThread(new Runnable() {
            public void run() {
                int rId = getResourceId(idName);
                if (rId > 0) {
                    String googleId = activity.getResources().getString(rId);
                    getGameHelper().getGamesClient().submitScore(googleId, score);
                }
            }
        });
    }

    public static void clearAllScores()
    {
        Log.v(TAG, "clearAllScores is not available on this platform");
    }

    /**
     *
     * Public interface to integrate this into the main activity
     *
     */
     
    public static void onStart()
    {
        getGameHelper().onStart(activity);
    }

    public static void onStop()
    {
        if (mHelper == null) {
            return;
        }

        getGameHelper().onStop();
    }

    public static void onActivityResult(int requestCode, int resultCode, Intent data)
    {
        if (mHelper == null) {
            return;
        }

        getGameHelper().onActivityResult(requestCode, resultCode, data);
    }

    /**
     *
     * Private helper methods
     *
     */

    private static int getResourceId(String idName)
    {
        int rId = activity.getResources().getIdentifier(idName, "string", activity.getPackageName());
        if (rId == 0) {
            Log.v(TAG, "Unable to find resource ID for string " + idName);
        }
        return rId;

    }

    private static GameHelper getGameHelper()
    {
        if (mHelper == null) {
            mHelper = new GameHelper(activity);
            mHelper.enableDebugLog(true, TAG);

            activity.runOnUiThread(new Runnable() {
                public void run() {
                    mHelper.setup(
                        new GameHelper.GameHelperListener() {
                            @Override
                            public void onSignInFailed() {
                                Log.v(TAG, "onSignInFailed");
                            }

                            @Override
                            public void onSignInSucceeded() {
                                Log.v(TAG, "onSignInSucceeded");
                            }
                        }, GameHelper.CLIENT_GAMES
                    );
                }
            });
        }

        return mHelper;
    }
}